[mkdocs] docs: add account metadata tutorial (#1803)
symbol/symbol: mkdocs/pages/en/devbook/accounts/account-metadata.md

---
title: Account Metadata
---

# Adding Metadata to an Account

<Accounts:> can store <metadata:> as key-value pairs.

This tutorial shows how to add metadata to an account, retrieve metadata from the network, and update existing values.

> In this example, an account adds a username to itself and then modifies it.

Reviewer: A `username` looks like a specific property of accounts and might be confusing. How about:

```suggestion
In this example, the string `alice` is attached an account and then changed to `bob`.
```

## Code Explanation

This tutorial demonstrates adding new metadata to an account and then updating that metadata.

### Setting Up the Account

{{ tutorial.code_snippet(['py:53:61', 'js:51:59']) }}

The snippet reads the signer's private key from the `SIGNER_PRIVATE_KEY` environment variable, which defaults to a
test key if not set.
The signer's address is derived from the public key.

In this tutorial, the signer adds metadata to their own account.
Adding metadata to a different account requires the target to cosign the transaction.

### Fetching Network Time and Fees

{{ tutorial.code_snippet(['py:64:82', 'js:62:80']) }}

Network time and recommended fees are fetched from <get:/node/time> and <get:/network/fees/transaction> respectively, following the process described in the [Transfer Transaction](./transfer.md) tutorial.

### Defining the Metadata

{{ tutorial.code_snippet(['py:84:87', 'js:82:85']) }}

Each metadata entry is identified by a **scoped metadata key**, a 64-bit value chosen by the metadata creator.
The combination of the signer's address, the target account's address, and the scoped metadata key uniquely

> identifies the metadata entry.

Reviewer: ```suggestion
Each metadata entry is uniquely identified by the signer's address, the target account's address, and a **scoped metadata key**: a 64-bit value chosen by the metadata creator.
```

!!! note "Multiple entries with the same key"

    Because the signer's address is part of the unique identifier, multiple accounts can add metadata with the same
    scoped key to a target (e.g., Account A and Account B can add a different `username` to Account C).

>     Each entry remains separate and can only be updated by its signer.

Reviewer: ```suggestion
    Because the signer's address is part of the unique identifier, different accounts can use the same scoped metadata key 
    on the same target account without conflict.
    
    For example, Account A and Account B can both use the key `username` when adding metadata to Account C,
    resulting in two distinct metadata entries.

    Each entry is independent and can only be updated by the account that originally created it.
```

The SDK provides a <dy:metadataGenerateKey> helper function that generates a key from a
human-readable string using SHA3-256 hashing.
This approach makes keys more meaningful and reduces the chance of collisions.

In this example, the key is derived from the string `username`.
To ensure the metadata key is unique across multiple runs of the tutorial, a timestamp is appended to the key string.
In practice, you would use a stable key that identifies the specific metadata entry you want to create or update.

The metadata value can be any byte sequence.

> In this example, the username `alice` is UTF-8 encoded.

Reviewer: ```suggestion
In this example, the value is the string `alice` encoded in UTF-8.
```

### Creating the Embedded Metadata Transaction

{{ tutorial.code_snippet(['py:89:101', 'js:87:100']) }}

> Symbol requires metadata transactions to be inside an <aggregate transaction:>.

Reviewer: It would be good to explain why.
I understand this is useful when the signer and the target accounts are different, but it looks cumbersome when they are not.
The Symbol Whitepaper does not mention metadata anywhere, so you'll have to ask in the chat.

> For this reason, the code defines the metadata update as an <embedded transaction:>.

Reviewer: You have not talked about updates yet, this section is about adding new metadata.

I know it's the same transaction, but the reader might not. So you better not talk about updates yet, or explain now that additions and updates are the same thing.
If you don't do it now you'll have to do it later.
Doing it now has the benefit that explaining below why Value and Value size delta are so weird is easier.

The account metadata transaction specifies:

* **Type:** Use `account_metadata_transaction_v1`.

* **Signer public key:** The account creating the metadata entry.

* **Target address:** The account to attach the metadata to.
  When the target differs from the signer, the target account must cosign the aggregate transaction.

> * **Scoped metadata key:** The 64-bit identifier for this metadata entry.

Reviewer: ```suggestion
* **Scoped metadata key:** The 64-bit key used to identify this metadata entry.
```

Previously we separated the key from the identifier that is obtained by combining the key with the addresses, so let's keep that nomenclature to avoid confusion.

* **Value size delta:** When creating new metadata, set this to the byte length of the value.
  When updating existing metadata, set this to the difference between the new and current value lengths.

* **Value:** The metadata content as bytes.
  When creating new metadata, provide the raw value.
  When updating, provide a computed value (explained in the
  [Modifying Existing Metadata](#modifying-existing-metadata) section).

### Building the Aggregate Transaction

{{ tutorial.code_snippet(['py:103:115', 'js:102:114']) }}

The code adds the embedded transaction to a complete <aggregate transaction:>.

Since the signer is modifying their own account, the transaction does not need cosignatures

> and the aggregate can be signed and announced immediately.

Reviewer: Mention that therefore this is an aggregate complete tx.

!!! note "Adding metadata to a different account"

    If the target account is different from the signer, the target must cosign the aggregate transaction to approve the
    metadata entry.
    This requirement prevents third parties from attaching unwanted metadata to an account without permission.

    For details on collecting <cosignatures:>, see the [Bonded Aggregate](../transactions/bonded-aggregate.md)
    tutorial.

### Announcing the Transaction

{{ tutorial.code_snippet(['py:117:125', 'js:116:125']) }}

The transaction is signed and announced following the same process as in
[Creating a Complete Aggregate Transaction](../transactions/complete-aggregate.md#building-the-aggregate-transaction).

### Retrieving Metadata

{{ tutorial.code_snippet(['py:130:147', 'js:130:147']) }}

Updating an existing metadata entry requires the current value from the network.

The code queries the <get:/metadata> endpoint with filters for `sourceAddress`, `targetAddress`,

> `scopedMetadataKey`, and `metadataType` (`0` for account metadata) to retrieve the specific entry.

Reviewer: Good that you mention that 0 is for account metadata, because the REST API docs say nothing about it 😩 

Are you interested in reviewing the REST API docs? 🙂

Author: It is described, but the Swagger UI does not render the enum description.

<img width="234" height="217" alt="image" src="https://github.com/user-attachments/assets/baa4d29a-49a7-46d3-b92a-a8aea587d408" />

I would definitely like to review the REST API docs. Let's add it to our list of follow-up items and discuss it in more detail later.

> ### Modifying Existing Metadata

Reviewer: Worth mentioning here that updating old metadata uses the same tx as adding new metadata, if you haven't explained it yet.

* **Line 16** (`"value": "616c696365"`): The metadata value encoded as hexadecimal (`"alice"` in UTF-8).
* **Line 27** (`"transactions_hash"`): The hash of the embedded transactions for the metadata creation.
* **Line 52** (`Current value: alice`): Retrieved from the network before updating.
* **Line 62** (`"transactions_hash"`): The hash of the embedded transactions for the metadata update.
* **Line 71** (`"value_size_delta": -2`): Negative because the new value (`"bob"` = 3 bytes) is shorter than the
  current value (5 bytes). The difference is -2.
* **Line 72** (`"value": "03030b6365"`): The XOR'd value computed from the current and new values, not the raw new
  value.

The transaction hashes printed in the confirmation messages can be used to search for the transactions
in the [Symbol Testnet Explorer](https://testnet.symbol.fyi/).

## Conclusion

This tutorial showed how to:

| Step                                                                                  | Related documentation                              |
| ------------------------------------------------------------------------------------- | -------------------------------------------------- |

> | [Define metadata key and value](#defining-the-metadata)                               | <dy:metadataGenerateKey>                           |

Author: @segfaultxavi `dy:metadataGenerateKey` and `dy:metadataUpdateValue` links do not seem to work. Could you please review?

Reviewer: These are global functions that were not handled by my `<dy:>` code, sorry!
It's fixed in the `new-docs` branch, you'll need to rebase and use `<dy:Metadata.metadataGenerateKey>`.

It's pretty hackish, but this will generate `<js:metadataGenerateKey>` in JS and `<py:Metadata.metadata_generate_key>` in Python.

---
title: Account Metadata
---

# Adding Metadata to an Account

<Accounts:> can store <metadata:> as key-value pairs.

This tutorial shows how to add metadata to an account, retrieve metadata from the network, and update existing values.

> In this example, the string `alice` is attached an account and then changed to `bob`.

Reviewer: ```suggestion
In this example, the string `alice` is attached to an account and then changed to `bob`.
```

Do not accept my suggestions without reviewing them :P

Additionally, review the [Transfer transaction](../transactions/transfer.md) tutorial to understand how
transactions are announced and confirmed, and the
[Complete Aggregate transaction](../transactions/complete-aggregate.md) tutorial to understand how aggregate
transactions work.

## Full Code

{% import 'tutorial.jinja2' as tutorial with context %}

{{ tutorial.code_full('devbook/accounts/account-metadata', ['py', 'js']) }}

## Code Explanation

This tutorial demonstrates adding new metadata to an account and then updating that metadata.

### Setting Up the Account

{{ tutorial.code_snippet(['py:53:61', 'js:51:59']) }}

The snippet reads the signer's private key from the `SIGNER_PRIVATE_KEY` environment variable, which defaults to a
test key if not set.
The signer's address is derived from the public key.

In this tutorial, the signer adds metadata to their own account.
Adding metadata to a different account requires the target to cosign the transaction.

### Fetching Network Time and Fees

{{ tutorial.code_snippet(['py:64:82', 'js:62:80']) }}

Network time and recommended fees are fetched from <get:/node/time> and <get:/network/fees/transaction> respectively,

> following the process described in the [Transfer Transaction](./transfer.md) tutorial.

Reviewer: ```suggestion
following the process described in the [Transfer Transaction](../transactions/transfer.md) tutorial.
```
There is a yellow warning in the mkdocs output log due to the missing file.

### Defining the Metadata

{{ tutorial.code_snippet(['py:87:90', 'js:85:88']) }}

Each metadata entry is uniquely identified by the signer's address, the target account's address, and a
**scoped metadata key**: a 64-bit value chosen by the metadata creator.

!!! note "Multiple entries with the same key"

    Because the signer's address is part of the unique identifier, different accounts can use the same scoped metadata
    key on the same target account without conflict.

    For example, Account A and Account B can both use the key `username` when adding metadata to Account C,
    resulting in two distinct metadata entries.

    Each entry is independent and can only be updated by the account that originally created it.

The SDK provides a <dy:Metadata.metadataGenerateKey> helper function that generates a key from a
human-readable string using SHA3-256 hashing.
This approach makes keys more meaningful and reduces the chance of collisions.

In this example, the key is derived from the string `username`.
To ensure the metadata key is unique across multiple runs of the tutorial, a timestamp is appended to the key string.

> In practice, you would use a stable key that identifies the specific metadata entry you want to create or update.

Reviewer: ```suggestion
For demonstration purposes, a timestamp is appended to the key string,
so each time the code is executed a new entry is added to the account.
In practice, you would use a fixed key that identifies the specific metadata entry you want to create or update.
```

I made it extra clear that the timestamp is only required for the tutorial and is not something users will normally need to do.

The metadata value can be any byte sequence.
In this example, the value is the string `alice` encoded in UTF-8.

### Creating the Embedded Account Metadata Transaction

{{ tutorial.code_snippet(['py:92:104', 'js:90:103']) }}

An account metadata transaction attaches a key-value pair to an account on the blockchain.

> The same transaction type handles both adding new metadata and updating existing entries.

Reviewer: ```suggestion
The same transaction type handles both adding new metadata entries and updating existing ones.
```

Otherwise it looks like "metadata" and "entries" are different things.

Symbol requires these transactions to be inside an <aggregate transaction:> to prevent third parties from attaching
unwanted metadata to an account without permission.
Even when target and signer are the same account, an aggregate is still required to keep the transaction format

> uniform.

Reviewer: ```suggestion
Symbol requires these transactions to be inside an <aggregate transaction:> that includes the target account owner's signature.  
This prevents unwanted metadata to be attached to an account without its owner's permission.  
An aggregate is still required even when the transaction is initiated by the account owner,
to keep the transaction format uniform.
```

Clarified.

For this reason, the code defines the account metadata transaction as an <embedded transaction:>.
This transaction specifies:

* **Type:** Use `account_metadata_transaction_v1`.

* **Signer public key:** The account creating the metadata entry.

> 

Reviewer: ```suggestion
    In this case, this is the account receiving the metadata too.

```

* **Target address:** The account to attach the metadata to.
  When the target differs from the signer, the target account must cosign the aggregate transaction.

* **Scoped metadata key:** The 64-bit key used to identify this metadata entry.

* **Value size delta:** When creating new metadata, set this to the byte length of the value.
  When updating existing metadata, set this to the difference between the new and current value lengths.

* **Value:** The metadata content as bytes.
  When creating new metadata, provide the raw value.
  When updating, provide a computed value (explained in the
  [Modifying Existing Metadata](#modifying-existing-metadata) section).

### Building the Aggregate Transaction

{{ tutorial.code_snippet(['py:106:116', 'js:105:115']) }}

> The code adds the embedded account metadata transaction to a <aggregate transaction:>.

Reviewer: ```suggestion
The code adds the embedded account metadata transaction to an <aggregate transaction:>.
```

Since the signer is modifying their own account, no <cosignatures:> are needed and the aggregate can be of type

> complete, which can be signed and announced immediately.

Reviewer: ```suggestion
Since the signer is modifying its own account, no <cosignatures:> are required and the aggregate can be created as
<complete aggregate transaction:|complete>, allowing it to be signed and announced immediately.
```

!!! note "Adding metadata to a different account"

    If the target account is different from the signer, the target must cosign the aggregate transaction to approve the
    metadata entry.

    For details on collecting cosignatures on-chain, see the [Bonded Aggregate](../transactions/bonded-aggregate.md)
    tutorial.

### Submitting the Aggregate Transaction

{{ tutorial.code_snippet(['py:118:127', 'js:117:128']) }}

The aggregate transaction is signed and announced following the same process as in
[Creating a Complete Aggregate Transaction](../transactions/complete-aggregate.md#building-the-aggregate-transaction).

### Retrieving Metadata

{{ tutorial.code_snippet(['py:132:149', 'js:133:150']) }}

Updating an existing metadata entry requires the current value from the network.

The code queries the <get:/metadata> endpoint with filters for `sourceAddress`, `targetAddress`,
`scopedMetadataKey`, and `metadataType` (`0` for account metadata) to retrieve the specific entry.

### Modifying Existing Metadata

{{ tutorial.code_snippet(['py:151:177', 'js:152:180']) }}

> To demonstrate updating metadata, the code changes the username from `alice` to `bob` using an embedded account

Reviewer: ```suggestion
To demonstrate updating metadata, the code changes the username from `alice` to `bob` using another account
```
